### ISSUE TITLE
Sometimes missing a call to HSET after job finishes?

### ISSUE THREAD

**Issue by KMouratidis** [rq/rq#1407]
tl;dr:
1. I submit a job. 
2. Job status is `queued`
3. A worker picks it up, set itselfs to `busy`, `current_job`, etc
4. Job status is still `queued`
5. The worker finishes, prints in its stdout a "Job OK", sets itself to `idle`, works fine
6. Job status is still `queued` 
7. ~9 minutes later it marks the job as `failed` :slightly_frowning_face: 

This happens only ~15 times over ~260 jobs, the remaining ~245 are all successful. I'm trying to find if this occurs under some rare condition I'm unaware of or if I've screwed up hard on my code, and maybe if you have some suggestions on how I could proceed with debugging. 

Setup: 
- 2 workers named `db_0` and `db_1` (managed by supervisord)
- Each worker listening to 3 queues  (`db:high`, `db:mid`, `db:low`)
- RQ: 1.7.0, just changed from 1.5.2
- Custom connection (see below), including pubsub implementation etc

I truncated UUIDs and a prefixes in logs so they are somewhat readable. 

Job enqueue:
```
>>> job.enqueued_at
datetime.datetime(2021, 1, 28, 11, 47, 37, 665461)
```

Worker `db_0` STDOUT:
```
06:47:38 db:low: tasks.do_something(args, kwargs) (db-db505d91064f)
06:47:38 db:low: Job OK (db-a793-db505d91064f)
06:47:38 Result is kept for 86400 seconds
```

Since I don't use Redis at all and wrote my own class with the same interface I have attached the logs of all the calls with their arguments that were made, only from `db_0`:
```
2021-01-28 06:56:50,875 -  blpop(['rq:queue:db:high', 'rq:queue:db:mid', 'rq:queue:db:low'], 15, )
2021-01-28 06:47:37,870 -  hgetall(b'rq:job:db-db505d91064f', )
2021-01-28 06:47:38,128 -  hset(rq:worker:db_0, state, busy, )
2021-01-28 06:47:38,146 -  pipeline(, )
2021-01-28 06:47:38,147 -  hset(rq:worker:db_0, state, busy, )
2021-01-28 06:47:38,160 -  hset(rq:worker:db_0, current_job, db-db505d91064f, )
2021-01-28 06:47:38,180 -  hset(rq:worker:db_0, last_heartbeat, 2021-01-28T11:47:38.180147Z, )
2021-01-28 06:47:38,215 -  zadd(rq:wip:db:low, {'db-db505d91064f': 1611834488}, )
2021-01-28 06:47:38,223 -  hset(b'rq:job:db-db505d91064f', mapping={'last_heartbeat': '2021-01-28T11:47:38.222761Z', 'st
atus': 'started', 'started_at': '2021-01-28T11:)
2021-01-28 06:47:38,679 -  hset(rq:worker:db_0, state, idle, )
2021-01-28 06:47:38,689 -  Worker db_0 is alive.
2021-01-28 06:47:38,689 -  expire(rq:worker:db_0, 90, )
2021-01-28 06:47:38,696 -  hset(rq:worker:db_0, last_heartbeat, 2021-01-28T11:47:38.696293Z, )
2021-01-28 06:47:38,731 -  expire(rq:worker:db_0, 90, )
2021-01-28 06:47:38,736 -  hset(rq:worker:db_0, last_heartbeat, 2021-01-28T11:47:38.736589Z, )

....
tons of heartbeats, blpops, etc
....

2021-01-28 06:56:50,559 -  hgetall(b'rq:job:db-db505d91064f', )
2021-01-28 06:56:50,768 -  hset(b'rq:job:db-db505d91064f', status, failed, )
2021-01-28 06:56:50,797 -  info(server, )
2021-01-28 06:56:50,798 -  hset(b'rq:job:db-db505d91064f', mapping={'created_at': '2021-01-28T11:47:37.642912Z', 'data':
 b'x\x9c\xcc\xbdY\xb3\xe4Z\x96\xe7U\xd55teUV\xd)
2021-01-28 06:56:50,815 -  pipeline(, )
2021-01-28 06:56:50,817 -  hset(b'rq:job:db-db505d91064f', mapping={'created_at': '2021-01-28T11:47:37.642912Z', 'data':
 b'x\x9c\xcc\xbdY\xb3\xe4Z\x96\xe7U\xd55teUV\xd)
2021-01-28 06:56:50,832 -  expire(b'rq:job:db-db505d91064f', 31536000, )
2021-01-28 06:56:50,840 -  expire(rq:job:db-db505d91064f:dependents, 31536000, )
2021-01-28 06:56:50,842 -  expire(rq:job::db-db505d91064f:dependencies, 31536000, )
2021-01-28 06:56:50,844 -  zadd(rq:failed:db:low, {'db-db505d91064f': 1643371010}, )
2021-01-28 06:56:50,854 -  zremrangebyscore(rq:wip:db:low, 0, 1611835010, )
2021-01-28 06:56:50,862 -  zremrangebyscore(rq:failed:db:low, 0, 1611835010, )
2021-01-28 06:56:50,867 -  smembers(rq:workers:db:low, )
2021-01-28 06:56:50,869 -  pipeline(, )
2021-01-28 06:56:50,869 -  exists(rq:worker:db_1, )
2021-01-28 06:56:50,873 -  exists(rq:worker:db_0, )
2021-01-28 06:56:50,875 -  blpop(['rq:queue:db:high', 'rq:queue:db:mid', 'rq:queue:db:low'], 15, )
```

**Comment by KMouratidis** [comment]
It seems I made a wrong assumption about threads being used in the new version. `PubSub.run_in_thread` is somehow causing a disconnect I'm not catching and that's what triggers the missing `HSET`. Sorry :disappointed:
